### PR TITLE
Fix: reload and merge easy-install pth file before save

### LINKED ISSUE #3126: [BUG] easy-install.pth when updated during installation is overwritten without the recently added changes
### setuptools version

setuptools==60.9.3

### Python version

CPython 3.8

### OS

Ubuntu 20.04

### Additional environment information

Used in combination/on top of a CMake project using sub-projects via cmake FetchContent module : https://cmake.org/cmake/help/git-stage/module/FetchContent.html#fetchcontent

### Description

Basically : my setup.py is calling cmake to build a cmake project with pybind11 extensions.
the main project includes others cmake (sub-)projects using FetchContent. I control everything in the picture.
I want to be able to install in editable mode the main project and I want all the sub-projects to be also installed along the way as dependencies/automatically (given they all contains pybind11 extensions), also in editable mode/the same way the main/top one. I so tried to put a `pip install -e` or (python setup.py develop) of the sub-project(s) as a cmake dedicated custom target(s) on the main project. it's well correctly called .. BUT : actually it's a pip install or setup.py develop which is called **during/inside** another one.. 

and the net effect is all is done correctly BUT the easy-install.pth file gets overwritten by the main pip install (after the inner one finish).. basically the directory added to it by the inner installation is lost after the outer one writes its resulting one..

does that ring a bell to someone ?

### Expected behavior

I would have expected that the outer/main pip install/setup.py develop read again the content of the easy-install.pth file (so to get a fresh content) before writing it.. and merging what's need to as necessary/correctly.

### How to Reproduce

Call pip install / setup.py develop inside another one.

### Output

can produce on demand.

### PR DESCRIPTION
**Pull request by gst**
## Summary of changes

Refactor `PthDistributions._load` into/with a second `_load_raw`. Re-use `_load` from `PthDistributions.save()`.

and thus use that in save() to reload/refresh the file content before saving. and merge the eventual new found items.

Closes #3126

### CODE REVIEW

**Comment by abravalheri** [comment]
Hi @gst thank you very much for the PR! Any help is very appreciated.

I probably lack the knowledge to properly review your PR, so I will leave that for the other maintainers.
But something that would really help is if you can add a test case capturing the failing behaviour you are trying to solve.

**Comment by gst** [comment]
Hi @abravalheri 

I have got this : 

```python
    def test_many_pth_distributions_merge_together(self, tmpdir):
        pth_path = tmpdir.join("file1.pth")
        pth1 = PthDistributions(pth_path)
        pth2 = PthDistributions(pth_path)
        new_path = tmpdir.join("foobar")
        new_path.mkdir()  # must exist
        pth2.paths.append(str(new_path))
        pth2.save()
        new_path_str = str(new_path)
        assert new_path_str not in pth1.paths
        pth1.save()
        assert new_path_str in pth1.paths, "the new_path should have been added by pth1 as well"
```

that I added in `setuptools/tests/test_easy_install.py` but I can't succeed to make it run right now..

I'm not used to run setuptools tests I guess ;)

**Comment by abravalheri** [comment]
Hi @gst.

Something that is useful to run is `flake8 setuptools` (`pytest-flake8` errors are quite cryptic, so you can solve that first before running the tests).

If you have `tox` installed, I would suggest running `tox -- -p no:cov` first (coverage will add a lot of noise) and then if everything passes just `tox`. You can also add `-x` and the specific test file if the tests are too slow in your machine (e.g. `tox -- -p no:cov -x setuptools/tests/test_easy_install.py`).

**Comment by abravalheri** [comment]
Could you add a little comment/note in this test explaining in which scenario the `PthDistributions` would be duplicated?

> ```python
>         pth1 = PthDistributions(pth_path)
>         pth2 = PthDistributions(pth_path)
> ```

(That is a way of signing for people changing this file in the future why this test should not be removed).

**Comment by abravalheri** [comment]
(By the way, thank you very much for working in a test case, that really helps in the PR and to better understand the problem).

**Comment by gst** [comment]
That's my pleasure :) 

but I cannot get the test to succeed still.. I'm doube checking..

but it's not easy with tox I find. any direct `python` call that I could pdb with ?

**Comment by abravalheri** [comment]
> but it's not easy with tox I find. any direct `python` command that I could pdb with ?

That is a tricky question.

I usually rely on [pytest's ability to open pdb on failure](https://docs.pytest.org/en/latest/how-to/failures.html?highlight=pdb#using-python-library-pdb-with-pytest), e.g.:

```bash
tox -- -p no:cov -x --pdb setuptools/tests/test_easy_install.py
```

If you want to avoid `tox` completely it is more complicated... I imagine you can create a virtualenv install `.[testing]` inside of it and then proceed running:

```bash
pytest -p no:cov -x --pdb setuptools/tests/test_easy_install.py
```

(I never tested this way of running the tests though).

**Comment by gst** [comment]
Hi, any comment on the test case and the change/fix(or enhancement?) ?

**Comment by abravalheri** [comment]
Hi @gst, I will have a look on the test case, thank you very much!

Unfortunately, I will have to leave further review to other maintainers because this topic is a bit out of the areas of setuptools I am familiar with.

I you need these changes in your workflow urgently, there are a few workarounds that you can use while this PR gets reviewed. For example, in your `pyproject.toml` you can use:

```toml
[build-system]
requires = [
    "setuptools @ git+https://github.com/gst/setuptools@fix-easy-install-pth-file-not-reloaded-before-save",
    ...
]
```

File: setuptools/command/easy_install.py [hunk 1]
```diff
@@ -1579,50 +1579,73 @@ def get_exe_prefixes(exe_filename):
...
+        return paths, dirty or (paths and saw_import)
 
-        if self.paths and not saw_import:
-            self.dirty = True  # ensure anything we touch has import wrappers
-        while self.paths and not self.paths[-1].strip():
-            self.paths.pop()
+    def _load(self):
+        if os.path.isfile(self.filename):
+            paths, dirty = self._load_raw()
+            # lights cleaning:
```
Comment (abravalheri) [inline setuptools/command/easy_install.py:1621]: I am not familiar with the expression in this comment, is `lights cleaning` a common dev expression?

File: setuptools/command/easy_install.py [hunk 2]
```diff
@@ -1579,50 +1579,73 @@ def get_exe_prefixes(exe_filename):
...
+            while paths and not paths[-1].strip():
```
Comment (abravalheri) [inline setuptools/command/easy_install.py:1622]: This while loop removing empty lines at the end, right?
If that is the case, isn't it better just to skip all the empty lines in `_load_raw`?

I see that you have a comment there saying that you keep empty lines for formatting. Does that help a lot or is it just something we can skip for the sake of simplicity?

File: setuptools/tests/test_easy_install.py [hunk 3]
```diff
@@ -317,6 +317,27 @@ def test_add_from_site_is_ignored(self):
         pth.add(PRDistribution(location))
         assert not pth.dirty
 
+    def test_many_pth_distributions_merge_together(self, tmpdir):
+        """makes sure that if the pth file is modified under the hood then a(ny) PthDistribution
```
Comment (abravalheri) [inline setuptools/tests/test_easy_install.py:321]:
```suggestion
        """Make sure that if the pth file is modified under the hood then PthDistribution
```

File: setuptools/tests/test_easy_install.py [hunk 4]
```diff
@@ -317,6 +317,27 @@ def test_add_from_site_is_ignored(self):
...
+        will refresh its content before saving. and merging as necessary what needs to be.
```
Comment (abravalheri) [inline setuptools/tests/test_easy_install.py:322]:
```suggestion
        will refresh its content before saving, merging contents when necessary.
        This is useful, for example, when relying on CMake FetchContent during builds.
```

File: setuptools/command/easy_install.py [hunk 5]
```diff
@@ -1579,50 +1579,73 @@ def get_exe_prefixes(exe_filename):
...
+            else:
+                last_paths.remove(path)
+        for path in self.paths[:]:
+            if path not in last_paths \
+                    and not path.startswith(('import ', 'from ', '#')):
+                absolute_path = os.path.join(self.basedir, path)
+                if not os.path.exists(absolute_path):
+                    self.paths.remove(path)
+                    log.info("removing now non-existent path %r", path)
+                    last_dirty = True
```
Comment (abravalheri) [inline setuptools/command/easy_install.py:1646]: This 2 loops look tricky to me. Is there anyway we can use `set` logic (e.g. [`set.intersection`](https://docs.python.org/3/library/stdtypes.html?highlight=intersection#frozenset.intersection) and [`set.difference`](https://docs.python.org/3/library/stdtypes.html?highlight=difference#frozenset.difference)) to detect new and removed paths?

**Review by abravalheri — COMMENTED** [review]
Thank you very much @gst for working on this!

I added just some general code review comments. I am unable assess how this change might impact on existing code, or if there is any undesired side-effect for other use cases that might be caused by this change. So I will leave this part of the review for other maintainers.

Please correct me here if I am wrong: I am assuming this change is necessary to make *editable installs* (i.e. `pip install -e`) work correctly, right? (In general we want to avoid changing `easy_install` since it is deprecated, unless it is necessary to make other parts of setuptools work correctly).

Finally, I would like to say that my comments are just suggestions and I am aware that people have different coding styles :)
I also understand perfectly if you want to wait until you have feedback from other maintainers before spending more energy and add other changes to this PR (to make sure your efforts are not going to be wasted).

**Comment by abravalheri** [comment]
Currently, the CI seems to be unhappy.

I recommend always running `flake8 setuptools` and `tox` (the full test suite) before pushing.

Of course you just run `pytest` locally while working on the changes (so you can have quick feedback), but it is good to run `tox` once you are finished to make sure everything is in a good shape (and avoid unexpected CI errors).

**Comment by gst** [comment]
@abravalheri I made the first changes you proposed.

I ran tox locally and everything succeed but 2 test cases .. not related.

I'm waiting anymore comments or review.

File: setuptools/command/easy_install.py (hunk 5 above)
Comment (abravalheri) [reply setuptools/command/easy_install.py:1649]: You could use `set` to find out which ones are removed/news and then `list.remove` then from the original variable?

Comment (gst) [reply setuptools/command/easy_install.py:1649]: looks a bit complicated compared to just looping over the refreshed view of the file and then after on the current one (minus the ones that have been already been processed). IMHO.

Comment (gst) [reply setuptools/command/easy_install.py:1649]: I mean: I made it to be simple ;)

Comment (gst) [reply setuptools/command/easy_install.py:1649]: the 2 loops are achieving 2 different things.

**Comment by abravalheri** [comment]
Thank you very much @gst, that is really appreciated.

Just a minor comment: `flake8` seems to be still unhappy 😅.

**Comment by gst** [comment]
@abravalheri that's flaked now ;)

**Comment by jaraco** [comment]
Overall, this change looks reasonable. I'm sorry the pytest-flake8 reporting is so bad (due to https://github.com/tholo/pytest-flake8/issues/81). I'll try to review this soon.

**Comment by jaraco** [comment]
I found https://github.com/tholo/pytest-flake8/issues/22, which shows some users having difficulty with pytest-flake8 getting it to honor the settings, so maybe that's what's happening.

**Comment by jaraco** [comment]
Nah. I'll deal with that separately.

**Comment by gst** [comment]
ping

**Pull request merged**